zerver/actions/users.py

```diff
@@ -94,6 +96,93 @@ def do_delete_user(user_profile: UserProfile, *, acting_user: Optional[UserProfi
         )
 
 
+def do_delete_user_preserving_messages(user_profile: UserProfile) -> None:
+    """
+    This is a version of do_delete_user which does not delete message content.
+
+    The code is a bit tricky, because we want to, at some point, call
+    user_profile.delete() to trigger cascade deletes to clean up the variety
+    of related objects - but we need to avoid the cascades affecting
+    messages sent by the user.
+    """
+    if user_profile.realm.is_zephyr_mirror_realm:
+        raise AssertionError("Deleting zephyr mirror users is not supported")
+
+    do_deactivate_user(user_profile, acting_user=None)
+
+    subscribed_huddle_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.HUDDLE
+        ).values_list("recipient_id", flat=True)
+    )
```

Comment:
Does this need to happen inside the transaction?

Reply:
I think it'd potentially be a lot cleaner if this used the same approach we take for Message, of rewriting items inside the transaction to the new temporary user ID, and then proceeding from there.

Reply:
Good point, that's cleaner - I've changed the handling of Subscriptions to do the same as for Messages.